### ISSUE TITLE
unattended upgrade removes modules

### ISSUE THREAD

**Issue by root360-StefanHeitmueller** [oerdnj/deb.sury.org#1555]
**Frequently asked questions**

- [x] I have read [Frequenty Asked Questions](https://github.com/oerdnj/deb.sury.org/wiki/Frequently-Asked-Questions)

**Describe the bug**
We are running unattended upgrades for all our systems and on monday, it decided to upgrade php packages and drop "old" versions of the package.

unattended upgrades log:

```
2021-02-22 11:51:39,783 INFO Packages that will be upgraded: bind9-host dnsutils libbind9-160 libdns-export1100 libdns1100 libirs160 libisc-export169 libisc169 libisccc160 libisccfg160 liblwres160 libssl1.0.0 libssl1.1 openssl php-apcu php-apcu-bc php-common php-igbinary php-redis php5.6-apcu php5.6-igbinary php5.6-redis php7.0-apcu php7.0-apcu-bc php7.0-bcmath php7.0-bz2 php7.0-cli php7.0-common php7.0-curl php7.0-fpm php7.0-gd php7.0-igbinary php7.0-imap php7.0-intl php7.0-json php7.0-mbstring php7.0-mcrypt php7.0-mysql php7.0-opcache php7.0-readline php7.0-redis php7.0-soap php7.0-sqlite3 php7.0-xml php7.0-zip php7.1-apcu php7.1-apcu-bc php7.1-igbinary php7.1-redis php7.2-apcu php7.2-apcu-bc php7.2-igbinary php7.2-redis php7.3-apcu php7.3-apcu-bc php7.3-igbinary php7.3-redis php7.4-apcu php7.4-apcu-bc php7.4-igbinary php7.4-redis php8.0-apcu php8.0-igbinary php8.0-redis tideways-php
2021-02-22 11:51:39,784 INFO Writing dpkg log to /var/log/unattended-upgrades/unattended-upgrades-dpkg.log
2021-02-22 11:53:40,224 INFO All upgrades installed
2021-02-22 11:53:41,117 DEBUG marking php7.4-redis for removal
2021-02-22 11:53:42,260 DEBUG marking php7.3-apcu for removal
2021-02-22 11:53:43,371 DEBUG marking php5.6-redis for removal
2021-02-22 11:53:44,485 DEBUG marking php7.3-redis for removal
2021-02-22 11:53:45,621 DEBUG marking php7.4-apcu for removal
2021-02-22 11:53:46,771 DEBUG marking php5.6-apcu for removal
2021-02-22 11:53:47,904 DEBUG marking php7.1-apcu for removal
2021-02-22 11:53:49,027 DEBUG marking php7.1-redis for removal
2021-02-22 11:53:50,151 DEBUG marking php7.2-apcu for removal
2021-02-22 11:53:51,297 DEBUG marking php7.0-apcu for removal
2021-02-22 11:53:52,491 DEBUG marking php7.0-redis for removal
2021-02-22 11:53:53,665 DEBUG marking php7.2-redis for removal
2021-02-22 11:53:54,788 INFO Packages that were successfully auto-removed: php5.6-apcu php5.6-redis php7.0-apcu php7.0-redis php7.1-apcu php7.1-redis php7.2-apcu php7.2-redis php7.3-apcu php7.3-redis php7.4-apcu php7.4-redis
```
e.g. php-redis has changed it's `Provides` section:

```
$ apt-cache show php-redis
Package: php-redis
Priority: optional
Section: php
Installed-Size: 14
Maintainer: Debian PHP PECL Maintainers <team+php-pecl@tracker.debian.org>
Architecture: amd64
Version: 5.3.2+4.3.0-7+ubuntu18.04.1+deb.sury.org+1
Suggests: redis-server
Depends: php8.0-redis, php-igbinary
Pre-Depends: php-common (>= 2:69~)
Filename: pool/main/p/php-redis/php-redis_5.3.2+4.3.0-7+ubuntu18.04.1+deb.sury.org+1_amd64.deb
Size: 5792
MD5sum: 8c2b9f25855a78c154b2489a1bc60436
SHA1: e625b29fac903325dacffcc28c19ceadc2c8e9c8
SHA256: b3dc10ea97798493023ef910428dfbfa54c9e5e73a76a25c7db8bbd896552c39
Description-en: PHP extension for interfacing with Redis
 This extension allows php applications to communicate with the Redis
 persistent key-value store.  The php-redis module provides an easy
 object oriented interface.
Description-md5: b8a3de89cf7f4311899f1f5751dbc60a

Package: php-redis
Priority: optional
Section: php
Installed-Size: 2349
Maintainer: Debian PHP PECL Maintainers <team+php-pecl@tracker.debian.org>
Architecture: amd64
Version: 4.1.1-1+ubuntu18.04.1+deb.sury.org+10
Suggests: redis-server
Provides: php5.6-redis, php7.0-redis, php7.1-redis, php7.2-redis, php7.3-redis
Depends: php-igbinary, php-common (>= 1:7.0+33~), phpapi-20180731 | phpapi-20170718 | phpapi-20160303 | phpapi-20151012 | phpapi-20131226, libc6 (>= 2.14)
Filename: pool/main/p/php-redis/php-redis_4.1.1-1+ubuntu18.04.1+deb.sury.org+10_amd64.deb
Size: 330664
MD5sum: c7fd31936602e6fb6645eb062b7f2b63
SHA1: 37da21f70b165346dc35d7043edfd311af478435
SHA256: 0efde826d10e6180531eb0c5dd9cc2b07f0926a0c9f4811e303b400a1bcff727
Description-en: PHP extension for interfacing with Redis
 This extension allows php applications to communicate with the Redis
 persistent key-value store.  The php-redis module provides an easy
 object oriented interface.
Description-md5: b8a3de89cf7f4311899f1f5751dbc60a
```

where the old version included the files for all versions, the new version only provides 8.0 (thus our systems need the "old" packages php<version>-<module> again, which was dropped with the version before.

This then leads to errors like this:

```
PHP Fatal error:  Uncaught Error: Class 'Redis' not found in ...
```

We observed the following modules at least:

* apcu
* redis
* memchached
* mongodb

If i understand the change in Provides and Meta-Naming, this should affect all meta-packages like php-<module>.

We were already running all packages according to https://www.patreon.com/posts/october-update-42636315, but had php-<module> and php<version>-<module> installed in parallel, as we were already running the old version of php-<module> which conflicts with php<version>-module (e.g. `Breaks: php-redis (<< 5.3.2+4.3.0-7+ubuntu18.04.1+deb.sury.org+1~)`). So our saltstack needed to update the php-<module> to latest version before installing php<version>-<module> or apt will complain.

**To Reproduce**

Hardly to reproduce, you need a system with old packages, and let UU perform the upgrade. (whitelist the repo: `Unattended-Upgrade::Allowed-Origins { "LP-PPA-ondrej-php:bionic";};`)

**Expected behavior**

Updated packages should not break the modules ;)

**Distribution (please complete the following information):**
 - OS: Ubuntu 20.04
 - Architecture: amd64
 - Repository: Ubuntu PPA

**Comment by oerdnj** [comment]
See https://www.patreon.com/posts/47793617

Also I strongly recommend not enabling autoremove from unattended upgrade, disable it with: `Unattended-Upgrade::Remove-Unused-Dependencies  "false";`

**Comment by oerdnj** [comment]
I'll keep this open for a while since it's like third duplicate issue now.

**Comment by root360-StefanHeitmueller** [comment]
Okay, thanks for the info. So if i am right with the latest update there is no need for the non-versioned module package anymore?

This way we could strip a lot of saltstack code from our setup.

**Comment by root360-StefanHeitmueller** [comment]
FTR, all issues concerning the same:

* https://github.com/oerdnj/deb.sury.org/issues/1554
* https://github.com/oerdnj/deb.sury.org/issues/1552
* https://github.com/oerdnj/deb.sury.org/issues/1551
* https://github.com/oerdnj/deb.sury.org/issues/1548

**Comment by oerdnj** [comment]
Yes, exactly. That was the reason for the change, so you can treat all the extensions same.

**Comment by oerdnj** [comment]
@root360-StefanHeitmueller I've copied the last two updates from Patreon to the GitHub Sponsors Updates. Can you confirm that you got the update?

**Comment by root360-StefanHeitmueller** [comment]
Yes, thank you! :rocket:

**Comment by root360-StefanHeitmueller** [comment]
> See https://www.patreon.com/posts/47793617
> 
> Also I strongly recommend not enabling autoremove from unattended upgrade, disable it with: `Unattended-Upgrade::Remove-Unused-Dependencies "false";`

Hm, just doublechecked, we're using this already. Turns out there's also `Unattended-Upgrade::Remove-New-Unused-Dependencies "false";` and some corner cases and nifty decisions inside unattended upgrade which led to the situation.

**Comment by fevangelou** [comment]
I personally never installed un-versioned modules as you mention in your Patreon posts (e.g. here https://www.patreon.com/posts/october-update-42636315 & here https://www.patreon.com/posts/47793617).

Ever since I've used your repos, I have always organized my installed modules very specifically. So it's always been a case of install phpX.Y-apcu, phpX.Y-memcached and so on. Never php-apcu or php-memcached.

If I'm reading this right, these modules should have never been uninstalled with the recent changes, since I always originally installed the versioned ones. Is that correct?

**Comment by oerdnj** [comment]
> If I'm reading this right, these modules should have never been uninstalled with the recent changes, since I always originally installed the versioned ones. Is that correct?

That’s outside of the bailiwick of the individual packages. This is driven by the internal state of the individual packages. That said - if you ever installed say `php7.2-apcu` and it pulled `php-apcu` via Provides before the packages were split in October last year than there’s no internal state for `php7.2-apcu` and you need to run `apt install php7.2-apcu` to set the internal state to *manually* installed.

**Comment by oerdnj** [comment]
> Is `apt-get autoremove` out of the question now?

Running `apt autoremove` without review was always out of question. But you can use it and if it tries to remove `<package>` you don’t want to be removed then just set the package state to manually installed with `apt install <package>`.

**Comment by fevangelou** [comment]
Thank you for the clarifications Ondřej, much appreciated.

**Comment by mfn** [comment]
May I suggest to the repo owner to maybe pin the issue so it appears on https://github.com/oerdnj/deb.sury.org/issues ?
![image](https://user-images.githubusercontent.com/87493/109990881-1b886280-7d0a-11eb-9565-0e0fa0545be4.png)

**Comment by oerdnj** [comment]
> May I suggest to the repo owner to maybe pin the issue

Good idea!

**Comment by robocoder** [comment]
Thanks for keeping this open. I was trying to figure out why our ansible play to install php8.0-apcu was also pulling in php7.4-apcu-bc.

**Comment by MarkusLange** [comment]
> Thanks for keeping this open. I was trying to figure out why our ansible play to install php8.0-apcu was also pulling in php7.4-apcu-bc.

Yes I have the same Problem "php-apcu-bc" (is an Recommended package) try to install all php7.4 stuff but there is no php8.0-apcu-bc
